### PR TITLE
[BUG] fix bug for vectorized compaction

### LINKED ISSUE #9604: [Bug] vectorized compaction failed
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Version

branch: master 
commit id: 953429e

### What's Wrong?

4 server, 1fe 3be deploy, 
I use tools/tpch-tools to load 100G data, after it finished load 1 hour later,  I check it and find that lineitem table failed to do cumulative compaction. Then I found error log in be.WARNING ,
```
W0516 13:37:26.612212 801433 compaction.cpp:252] row_num does not match between cumulative input and output! input_row_num=18742298, merged_row_num=0, filtered_row_num=0, output_row_num=0
W0516 13:37:26.612402 801433 tablet.cpp:1410] failed to do cumulative compaction. res=Internal error: :  0# doris::Status::ConstructErrorStatus(short, doris::Slice const&) at /mnt/hdd01/ldy/incubator-doris/be/src/common/status.cpp:80
 1# doris::Compaction::check_correctness(doris::Merger::Statistics const&) at /mnt/hdd01/ldy/incubator-doris/be/src/olap/compaction.cpp:276
 2# doris::Compaction::do_compaction_impl(long) at /mnt/hdd01/ldy/incubator-doris/be/src/olap/compaction.cpp:121
 3# doris::Compaction::do_compaction(long) at /mnt/hdd01/ldy/incubator-doris/be/src/olap/compaction.cpp:57
 4# doris::CumulativeCompaction::execute_compact_impl() at /mnt/hdd01/ldy/incubator-doris/be/src/olap/cumulative_compaction.cpp:75
 5# doris::Compaction::execute_compact() at /mnt/hdd01/ldy/incubator-doris/be/src/olap/compaction.cpp:46
 6# doris::Tablet::execute_compaction(doris::CompactionType) at /mnt/hdd01/ldy/incubator-doris/be/src/olap/tablet.cpp:1407
 7# std::_Function_handler<void (), doris::StorageEngine::_submit_compaction_task(std::shared_ptr<doris::Tablet>, doris::CompactionType)::{lambda()#1}>::_M_invoke(std::_Any_data const&) at /var/local/ldb-toolchain/include/c++/11/bits/std_function.h:291
 8# doris::ThreadPool::dispatch_thread() at /mnt/hdd01/ldy/incubator-doris/be/src/util/threadpool.cpp:548
 9# doris::Thread::supervise_thread(void*) at /mnt/hdd01/ldy/incubator-doris/be/src/util/thread.cpp:409
10# start_thread in /lib/x86_64-linux-gnu/libpthread.so.0
11# __clone in /lib/x86_64-linux-gnu/libc.so.6
 (error -224), tablet=1007524.1617673526.414fcdf78976e781-857a7348cc73c68e
```
This error log still keep printing after I drop lineitem and drop database.
then I set enable_vectorized_compaction=false in be.conf and restart cluster, it stops print.
At this time, it looks like recovered, so I redo load again, and no error log again.

To summarize：
  Load tpch 100G data with enable_vectorized_compaction=true, NOT OK;
  Load tpch 100G data with enable_vectorized_compaction=false,  OK;

### What You Expected?

do cumulative compaction and base compaction successful as expected.

### How to Reproduce?

use tools/tpch-tools to load 100G data.

### Anything Else?

If more information is required, I will provide it

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by Gabriel39**
# Proposed changes

Issue Number: close #9604 

## Problem Summary:

Describe the overview of changes.

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
3. Has document been added or modified: (Yes/No/No Need)
4. Does it need to update dependencies: (Yes/No)
5. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/olap/vcollect_iterator.cpp [hunk 1]
```diff
@@ -55,7 +55,8 @@ void VCollectIterator::build_heap(std::vector<RowsetReaderSharedPtr>& rs_readers
         DCHECK(!rs_readers.empty());
         for (auto [c_iter, r_iter] = std::pair {_children.begin(), rs_readers.begin()};
              c_iter != _children.end();) {
-            if ((*c_iter)->init() != Status::OK()) {
+            auto s = (*c_iter)->init();
+            if (!s.ok() && s.precise_code() != OLAP_ERR_DATA_EOF) {
                 delete (*c_iter);
```
Comment (yiguolei) [inline be/src/vec/olap/vcollect_iterator.cpp:60]: Need to return an error status here?

Comment (Gabriel39) [reply be/src/vec/olap/vcollect_iterator.cpp:60]: Done

**Review by yiguolei — DISMISSED** [review]
LGTM

File: be/src/vec/olap/vcollect_iterator.cpp [hunk 2]
```diff
@@ -91,6 +103,7 @@ void VCollectIterator::build_heap(std::vector<RowsetReaderSharedPtr>& rs_readers
             }
             Level1Iterator* cumu_iter = new Level1Iterator(cumu_children, _reader,
                                                            cumu_children.size() > 1, _skip_same);
+            RETURN_IF_NOT_EOF_AND_OK(cumu_iter->init());
             cumu_iter->init();
```
Comment (Gabriel39) [inline be/src/vec/olap/vcollect_iterator.cpp:107]:
```suggestion
```

File: be/src/vec/olap/vcollect_iterator.cpp [hunk 3]
```diff
@@ -91,6 +103,7 @@ void VCollectIterator::build_heap(std::vector<RowsetReaderSharedPtr>& rs_readers
...
             std::list<LevelIterator*> children;
```
[the remark this answers is not in the document]
Comment (Gabriel39) [reply be/src/vec/olap/vcollect_iterator.cpp:107]: Forgot to remove this line and I did this just now

**Review by yiguolei — APPROVED** [review]
LGTM

**Pull request merged**